test/distributed/fsdp/test_fsdp_summon_full_params.py

```diff
@@ -0,0 +1,133 @@
+# Owner(s): ["oncall: distributed"]
+
+import functools
+import sys
+import math
+
+import torch
+import torch.nn as nn
+from torch import distributed as dist
+from torch.distributed._fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed._fsdp.wrap import default_auto_wrap_policy
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import (
+    FSDPTest,
+)
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    run_tests,
+)
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestSummonFullParams(FSDPTest):
+    @property
+    def world_size(self):
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_no_writeback(self):
+        model = FSDP(nn.Linear(4, 4).cuda(self.rank))
+        # set the value
+        p = list(model.parameters())[0]
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with model._summon_full_params(volatile=True):
+            with torch.no_grad():
+                p.copy_(torch.zeros_like(p))
+
+        self.assertEqual(p[0], self.rank + 2)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_with_writeback(self):
+        model = FSDP(nn.Linear(3, 3).cuda(self.rank))
+        # set the value
+        p = list(model.parameters())[0]
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with torch.no_grad():
+            with model._summon_full_params(volatile=False):
+                p.copy_(torch.zeros_like(p))
+
+        self.assertEqual(p[0], 0)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_shard_value(self):
+
+        raw_model = nn.Linear(10, 11)
+        # params: 11x10 + 11 -> 121
+        self.assertEqual(121, sum([m.numel() for m in raw_model.parameters()]))
+
+        model = FSDP(raw_model.cuda(self.rank))
+
+        # params should be half as many
```

Comment:
Is it more accurate to say params should be 121 / world size as we can have world size > 2?

Reply:
yup. I changed the test to be world_size agnostic.